File: game/tv_show/tv_show_parent.py
import random

from config import Config
from game_data import GameData
from phone_events import PhoneEvents
from tv_show.attract import Attract
from tv_show.playing import Playing


class TvShowParent:
    def __init__(self, context: GameData):
        self.context = context
        self._state = Attract(context)
        self._state.on_enter()
        self.country = context.country
        self.current_game = next(iter(Config.GAMES))

    def set_random_game(self):
        if len(Config.GAMES) == 1:
            self.current_game = next(iter(Config.GAMES))
            return
        new_game = random.choice(list(Config.GAMES.keys()))
        while new_game == self.current_game:
            new_game = random.choice(list(Config.GAMES.keys()))
        self.current_game = new_game

    def handle_events(self, phone_events: PhoneEvents):
        next_state = self._state.process_events(phone_events)

        if next_state is not None:
            self._state.on_exit()
            self._state = next_state(self.context)
            self._state.on_enter()

    def render(self, screen):
        self._state.render(screen)

    def is_playing(self):
        return not isinstance(self._state, Attract)

    def external_effect(self, effect):
        if isinstance(self._state, Playing):
            self._state.external_effect(effect)

File: game-web/src/tvShow/tvShowParent.ts
/** Port of game/tv_show/tv_show_parent.py */

import { Config } from '../config';
import { Attract } from './attract';
import { Playing } from './playing';
import type { EffectType } from '../effectType';
import type { GameData } from '../gameData';
import type { PhoneEvents } from '../phoneEvents';
import type { State } from '../state';

export class TvShowParent {
  readonly country: string;
  currentGame: string;

  private context: GameData;
  private state: State;

  constructor(context: GameData) {
    this.context = context;
    this.country = context.country;
    this.currentGame = Object.keys(Config.GAMES)[0];

    this.state = new Attract(context);
    this.state.onEnter();
  }

  setRandomGame(): void {
    const games = Object.keys(Config.GAMES);
    if (games.length === 1) {
      this.currentGame = games[0];
      return;
    }
    let newGame: string;
    do {
      newGame = games[Math.floor(Math.random() * games.length)];
    } while (newGame === this.currentGame);
    this.currentGame = newGame;
  }

  handleEvents(phoneEvents: PhoneEvents): void {
    const nextStateClass = this.state.processEvents(phoneEvents);
    if (nextStateClass !== null) {
      this.state.onExit();
      this.state = new nextStateClass(this.context);
      this.state.onEnter();
    }
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    this.state.render(ctx);
  }

  isPlaying(): boolean {
    return !(this.state instanceof Attract);
  }

  externalEffect(effect: EffectType): void {
    if (this.state instanceof Playing) {
      this.state.externalEffect(effect);
    }
  }

  cleanup(): void {
    this.state.onExit();
  }
}
